Fix the 'All Traders Open' feature

K4rian opened the pull request:
This PR addresses a bug where traders' doors remain closed in the first game after a server launch when `bAllTradersOpen=true`.
Not a major issue since restarting or changing the map resolves it, but it can still be confusing.

**Steps to Reproduce:**
- Open `KFPatcherSettings.ini` and set `bAllTradersOpen=true`.
- Start a vanilla server with `KFPatcher` as `ServerActors`.
- Beat the first wave, only the usual selected trader is accessible.

**Notes:**
- The fix is applied only once per server lifespan, during the first game's initial wave.
- The second `for loop` might be overkill, but it ensures that all valid shops are in the shop list.
- The fix has been tested for a couple of days without issue.

Thanks.

K4rian committed:
Fix 'All Traders Open' not applying in the first game after server launch

K4rian committed:
Match project style formatting

Shtoyan commented:
Hi. Thanks for your contribution!

There is a little issue with your approach - you should not create new local variables. KFPatcher is not a usual KF1 mod, it uses few loopholes in unrealscript and has several limitations. Unless you want to get random crashes at random time.

I'll try to reproduce this issue and see how else we can fix it.

K4rian commented:
Hey,

You're absolutely right, I reviewed the notes on the Unflect project about local variables, I'll update my code ASAP. Everything else seems to be working as intended (dosh throwing limits, trader messages, etc.).

Thanks for your review.

**EDIT**: I've merged the trader registration logic into a static function in Utility to eliminate code duplication and remove the local variables.

K4rian committed:
Add RegisterAllTraders static function in Utility to handle trader registration

Shtoyan commented on Classes/hookGT.uc:
```diff
@@ -21,6 +21,10 @@ var bool bBossView;
 var float BossViewBackTime;
 var transient ZombieBoss BossArray;
 
+// all traders open fix
+// only applied once during the server's lifespan
+var bool bAllTradersOpenFixApplied;
+
```
We don't need this.

K4rian replied:
Removed.
Without this variable, the `RegisterAllTraders` function will be called repeatedly.

Shtoyan commented on Classes/hookGT.uc:
```diff
@@ -526,6 +513,13 @@ state MatchInProgress
 
     else if ( NumMonsters <= 0 )
     {
+      // apply the 'All Traders Open' fix only during the initial wave, if enabled
```
Also this is not required. Just checked it in game.

K4rian replied:
The comment has been removed.

Shtoyan commented on Classes/hookGT.uc:
```diff
@@ -49,25 +53,8 @@ event InitGame(string Options, out string Error)
     else Warn("MULTIPLE KFLEVELRULES FOUND!!!!!");
   }
 
-  // added all traders open option
```
And I can't comment the `Utility` class changes, so typing here:
Nice idea, but I didn't get why you added `bSVExists` variable and why you manually assign ShopVolume URL's. Just keep the original checks / logic.

K4rian replied:
The `SVExists` variable is used to check whether a Shop Volume already exists in the `ShopList`. This check is done using the `URL` field as a comparison, not an assignment.

Without this check, the `ShopList` array will get bloated with duplicates. Unless you move the `RegisterAllTraders` call in `hookGT` somewhere else, it's best to keep this check.

Feel free to remove the logic on your side and add a log message to see what happens.

**Note:** No visible changes because the `Utility.uc` file was saved with `UTF-16 LE` encoding. I'm guessing you created or edited it on Windows? It should be `UTF-8`, like all the other files, or Git will treat it as a binary file instead of text.

Shtoyan commented on Classes/hookGT.uc:
```diff
@@ -49,25 +53,8 @@ event InitGame(string Options, out string Error)
+  // add traders
+  class'Utility'.static.RegisterAllTraders(self, ShopList, bUsingObjectiveMode);
 
   if (class'Settings'.default.bAllTradersOpen)
```
This can also be moved into `Utility`, to make everything clean.

K4rian replied:
Your log message can't be moved to `RegisterAllTraders`, or it will spam the console.

Shtoyan reviewed (changes requested):
Added few notes.

K4rian committed:
Remove bAllTradersOpenFixApplied

- hookGT: Removed the `bAllTradersOpenFixApplied` class variable and updated the corresponding `if` condition
- Utility (RegisterAllTraders): Restored a None check for the shop volume in the for loop

K4rian commented:
I made some of the changes you asked for.

Were you able to reproduce the bug on your side, like mentioned in the original PR description?

Shtoyan committed:
Fix `utility`'s encoding

Don't know why it was non utf8 at first place.

Shtoyan committed:
All Traders open feature touchup

Added new function hook, so code runs only once.

Shtoyan commented:
@K4rian pushed the changed logic. Don't forget to update your config, and check if it's ok.

K4rian committed:
Formatting touchup

K4rian commented:
Brilliant. It works like a charm, thanks!
I've made some final touchups, to preserve the original code formatting.
